docs/concepts/filters.md

```diff
@@ -0,0 +1,307 @@
+    [Authorize(Roles = "Admin")] // Requires Admin role
+    public static string ManageWeatherAlerts(string alertType)
+    {
+        return $"Managing alert: {alertType}";
+    }
+}
+```
+
+### Class-Level Authorization
+
+You can apply authorization at the class level, which affects all tools in the class:
+
+```csharp
+[McpServerToolType]
+[Authorize] // All tools require authentication
+public class RestrictedTools
+{
+    [McpServerTool, Description("Restricted tool accessible to authenticated users")]
+    public static string RestrictedOperation()
+    {
+        return "Restricted operation completed";
+    }
+
+    [McpServerTool, Description("Public tool accessible to anonymous users")]
+    [AllowAnonymous] // Overrides class-level [Authorize]
+    public static string PublicOperation()
+    {
+        return "Public operation completed";
+    }
+}
+```
+
+### How Authorization Filters Work
+
+The authorization filters work differently for list operations versus individual operations:
+
+#### List Operations (ListTools, ListPrompts, ListResources)
+For list operations, the filters automatically remove unauthorized items from the results. Users only see tools, prompts, or resources they have permission to access.
+
+#### Individual Operations (CallTool, GetPrompt, ReadResource)
+For individual operations, the filters return authorization errors when access is denied:
+
+- **Tools**: Returns a `CallToolResult` with `IsError = true` and an error message
+- **Prompts**: Throws an `McpException` with "Access forbidden" message
+- **Resources**: Throws an `McpException` with "Access forbidden" message
+
+### Filter Execution Order and Authorization
+
+Authorization filters are applied automatically when you call `AddAuthorizationFilters()`. These filters run at a specific point in the filter pipeline, which means:
+
+**Filters added before authorization filters** can see:
+- Unauthorized requests for operations before they are rejected by the authorization filters
+- Complete listings for unauthorized primitives before they are filtered out by the authorization filters
+
+**Filters added after authorization filters** will only see:
+- Authorized requests that passed authorization checks
+- Filtered listings containing only authorized primitives
+
+This allows you to implement logging, metrics, or other cross-cutting concerns that need to see all requests, while still maintaining proper authorization:
+
+```csharp
+services.AddMcpServer()
+    .WithHttpTransport()
+    .AddListToolsFilter(next => async (context, cancellationToken) =>
+    {
+        // This filter runs BEFORE authorization - sees all tools
+        Console.WriteLine("Request for tools list - will see all tools");
+        var result = await next(context, cancellationToken);
+        Console.WriteLine($"Returning {result.Tools?.Count ?? 0} tools after authorization");
+        return result;
+    })
+    .AddAuthorizationFilters() // Authorization filtering happens here
```

Comment:
I was wondering why we required this to be called explicitly, but it seems it's so that the order can be controlled? Do we want to behave as if it's automatically called if the user never calls it? Or are there scenarios where someone would legitimately use attributes / register filters, but not want this method called?

Reply:
The original version of this PR added it automatically in `WithHttpTransport`, but I thought it was very odd that the order of the authorization filter was dependent on the order you called `WithHttpTransport` vs the other filters.

I don't think there's a scenario where someone would legitimately use attributes / register filters, but not want this method called which is why I added the IPostConfigureOptions check that throws when this is the case. We could have IPostConfigureOptions just add the authorization filters at the end instead of throw, but I like being explicit about the order. We started adding `UseAuthorization` automatically in `WebApplicationBuilder` around .NET 7, and it's lead to a lot of confusion.